src/server/pps/server/pipeline_controller.go

```diff
@@ -191,19 +191,64 @@ func (pc *pipelineController) step(timestamp time.Time) (retErr error) {
 		tracing.FinishAnySpan(span, "err", retErr)
 	}()
 
+	if step, ctx, err := pc.newStep(); err != nil {
```

Comment:
since both of the first two branches return, I'd slightly prefer having the last big block indented one level less